rsocket-ipc-core/src/main/java/io/rsocket/ipc/IPCServerRSocket.java

```diff
@@ -0,0 +1,181 @@
+      String method = Metadata.getMethod(metadata);
+
+      BiFunction<Object, ByteBuf, Mono<Void>> ff = this.ff.get(method);
+
+      if (ff == null) {
+        return Mono.error(
+            new NullPointerException("nothing found for service " + service + " method " + method));
+      }
+
+      Object input = unmarshaller.apply(data);
+
+      return ff.apply(input, buf);
+
+    } catch (Throwable t) {
+      return Mono.error(t);
+    } finally {
+      payload.release();
+    }
+  }
+
+  @Override
+  public Mono<Payload> requestResponse(Payload payload) {
+    try {
+      ByteBuf data = payload.sliceData();
+      ByteBuf metadata = payload.sliceMetadata();
+
+      ByteBuf buf = Metadata.getMetadata(metadata);
+      String method = Metadata.getMethod(metadata);
+      BiFunction<Object, ByteBuf, Mono> rr = this.rr.get(method);
+
+      if (rr == null) {
+        return Mono.error(
+            new NullPointerException("nothing found for service " + service + " method " + method));
+      }
+
+      Object input = unmarshaller.apply(data);
+
+      return rr.apply(input, buf).map(this::marshall);
+
+    } catch (Throwable t) {
+      return Mono.error(t);
+    } finally {
+      payload.release();
+    }
+  }
+
+  @Override
+  public Flux<Payload> requestStream(Payload payload) {
+    try {
+      ByteBuf data = payload.sliceData();
+      ByteBuf metadata = payload.sliceMetadata();
+
+      ByteBuf buf = Metadata.getMetadata(metadata);
+      String method = Metadata.getMethod(metadata);
+
+      Object input = unmarshaller.apply(data);
+      BiFunction<Object, ByteBuf, Flux> rs = this.rs.get(method);
+
+      if (rs == null) {
+        return Flux.error(
+            new NullPointerException("nothing found for service " + service + " method " + method));
+      }
+
+      return rs.apply(input, buf).map(this::marshall);
+
+    } catch (Throwable t) {
+      return Flux.error(t);
+    } finally {
+      payload.release();
+    }
+  }
+
+  @Override
+  public Flux<Payload> requestChannel(Payload payload, Flux<Payload> publisher) {
+    try {
+      ByteBuf data = payload.sliceData();
+      ByteBuf metadata = payload.sliceMetadata();
+
+      ByteBuf buf = Metadata.getMetadata(metadata);
+      String method = Metadata.getMethod(metadata);
+
+      Object input = unmarshaller.apply(data);
+      TriFunction<Object, Publisher, ByteBuf, Flux> rc = this.rc.get(method);
+
+      if (rc == null) {
+        return Flux.error(
+            new NullPointerException("nothing found for service " + service + " method " + method));
+      }
+
+      Flux f =
+          publisher.map(
+              p -> {
+                try {
+                  Object o = unmarshaller.apply(p.sliceData());
+                  return o;
+                } finally {
+                  p.release();
+                }
+              });
+
+      return rc.apply(input, f, buf).map(this::marshall);
+
+    } catch (Throwable t) {
+      return Flux.error(t);
+    }
```

Comment:
Is the reason for not releasing payload here, that it's also the first payload in the stream? I forget where we landed on this?